src/Servers/Kestrel/Core/src/Internal/Http2/Http2FrameWriter.cs

```diff
@@ -286,6 +286,44 @@ public ValueTask<FlushResult> WriteDataAsync(int streamId, StreamOutputFlowContr
             }
         }
 
+        public ValueTask<FlushResult> WriteDataAndTrailersAsync(Http2Stream stream, StreamOutputFlowControl flowControl, in ReadOnlySequence<byte> data, bool firstWrite)
+        {
+            // The Length property of a ReadOnlySequence can be expensive, so we cache the value.
+            var dataLength = data.Length;
+
+            lock (_writeLock)
+            {
+                if (_completed || flowControl.IsAborted)
+                {
+                    return default;
+                }
+
+                // Zero-length data frames are allowed to be sent immediately even if there is no space available in the flow control window.
+                // https://httpwg.org/specs/rfc7540.html#rfc.section.6.9.1
+                if (dataLength != 0 && dataLength > flowControl.Available)
```

Comment:
I see this code is pretty difficult to share since we want to take different execution paths based on flow control availability and not pipe back pressure, and we wouldn't be able to tell which is which bu just calling WriteDataAsync() directly and checking the task. That said, I'm a bit worried that in the future we'll make a change in WriteDataAsync() that isn't reflected here. At the very least we should add a scary comment in WriteDataAsync() telling anyone editing that method to look at this one too.

Reply:
Add a comment in WriteDataAsync